### PR TITLE
DrivAerML: sweep cosine_t_max (10/20/30/50) for epoch-budget fit

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

With the default cosine_t_max=150 and only 5-15 epochs completed in the DrivAerML budget, the LR never meaningfully anneals. A shorter cosine cycle (cosine_t_max=10–30) would complete 1-3 full annealing cycles and might accelerate convergence. We sweep 4 values to find the best schedule for DrivAerML's epoch budget.

## Current Baseline

- DrivAerML: **no baseline yet** — target to beat: surface_rel_l2_pct < 3.71% (Transolver-3)

## Instructions

Run 4 parallel trials with `CUDA_VISIBLE_DEVICES`. Use `--wandb_group radford-drivaerml-round1` for all runs. AdamW lr=5e-4, batch_size=1.

### Trial 0 — GPU 0: cosine_t_max=10
```bash
CUDA_VISIBLE_DEVICES=0 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine_t_max 10 --batch-size 1 --wandb_group radford-drivaerml-round1 --wandb_name taki/drivaerml-tmx10 --agent taki
```

### Trial 1 — GPU 1: cosine_t_max=20
```bash
CUDA_VISIBLE_DEVICES=1 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine_t_max 20 --batch-size 1 --wandb_group radford-drivaerml-round1 --wandb_name taki/drivaerml-tmx20 --agent taki
```

### Trial 2 — GPU 2: cosine_t_max=30
```bash
CUDA_VISIBLE_DEVICES=2 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine_t_max 30 --batch-size 1 --wandb_group radford-drivaerml-round1 --wandb_name taki/drivaerml-tmx30 --agent taki
```

### Trial 3 — GPU 3: cosine_t_max=50
```bash
CUDA_VISIBLE_DEVICES=3 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine_t_max 50 --batch-size 1 --wandb_group radford-drivaerml-round1 --wandb_name taki/drivaerml-tmx50 --agent taki
```

Run all 4 in parallel (background each with `&`). Wait for all to finish.

**Notes:**
- Primary metric: `val_primary/surface_rel_l2_pct` (lower is better)
- Report epochs completed and the best val_primary/surface_rel_l2_pct for each trial
- Note the LR at final epoch for each trial (shows where in the cosine cycle we end)

## Results

_To be filled by student after runs complete._

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT taki:

## Results

### Summary

| Trial | cosine_t_max | val_primary/surface_rel_l2_pct | test_primary/surface_rel_l2_pct | train_loss | final LR | train_steps |
|-------|-------------|-------------------------------|--------------------------------|-----------|----------|-------------|
| **tmx50** | **50** | **36.05%** | **34.62%** | **0.395** | 3.97e-4 | 3585 |
| tmx10 | 10 | 37.07% | 35.62% | 0.408 | 2.50e-4 | 3585 |
| tmx20 | 20 | 38.59% | 37.22% | 0.415 | 7.32e-5 | 3585 |
| tmx30 | 30 | 41.20% | 39.80% | 0.450 | 2.50e-4 | 3585 |

**Winner: cosine_t_max=50** with 36.05% val / 34.62% test surface_rel_l2_pct.

### Commands used

```bash
# All trials: AdamW lr=5e-4, batch_size=1, 1M surface points
# Note: --drivaerml-train-surface-points 1000000 was required to avoid OOM
# (default 0 = all 8.8M points per case → OOM in Transolver forward pass)

CUDA_VISIBLE_DEVICES=0 python -u train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 10 --batch-size 1 --drivaerml-train-surface-points 1000000 --drivaerml-eval-surface-points 1000000 --wandb-group radford-drivaerml-round1 --wandb-name "taki/drivaerml-tmx10" --agent taki
CUDA_VISIBLE_DEVICES=1 python -u train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 20 --batch-size 1 --drivaerml-train-surface-points 1000000 --drivaerml-eval-surface-points 1000000 --wandb-group radford-drivaerml-round1 --wandb-name "taki/drivaerml-tmx20" --agent taki
CUDA_VISIBLE_DEVICES=2 python -u train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --batch-size 1 --drivaerml-train-surface-points 1000000 --drivaerml-eval-surface-points 1000000 --wandb-group radford-drivaerml-round1 --wandb-name "taki/drivaerml-tmx30" --agent taki
CUDA_VISIBLE_DEVICES=3 python -u train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 50 --batch-size 1 --drivaerml-train-surface-points 1000000 --drivaerml-eval-surface-points 1000000 --wandb-group radford-drivaerml-round1 --wandb-name "taki/drivaerml-tmx50" --agent taki
```

### Peak memory usage
~57.7 GB per GPU (H100 95GB)

### W&B run IDs
- tmx10: `srhwteb2` — https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/srhwteb2
- tmx20: `qauq34bl` — https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/qauq34bl
- tmx30: `yb8gw9kt` — https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/yb8gw9kt
- tmx50: `mxymmatz` — https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/mxymmatz

### What happened

**cosine_t_max=50 is the clear winner**, outperforming all other values on both val and test. The ranking is 50 > 10 > 20 > 30 — not monotonic, suggesting the optimal cycle length is neither too short nor too long.

**Key observation: cosine_t_max is per-step, not per-epoch.** The `scheduler.step()` is called after every batch, so with 3585 training steps per epoch:
- tmx10: 358 full cosine cycles per epoch (very rapid oscillation)
- tmx20: 179 cycles
- tmx30: 119 cycles  
- tmx50: 71 cycles

The longer cycle (tmx50) gives the optimizer more time at near-peak LR before annealing, which appears beneficial. The shorter cycles (tmx10, 20) oscillate too rapidly, reducing effective learning rate on average. tmx30 performed worst — possibly hitting an unlucky resonance with the data distribution.

**Important DrivAerML note:** The advisor's original commands lacked `--drivaerml-train-surface-points`, which defaults to 0 (all ~8.8M surface points per case). This OOMs because the senpai_transolver model processes surface_x directly (unlike reference_abupt which uses ABUPTCollate to subsample to 25k geometry points). I tested 50k through 4M surface points:
- 1M: works (57.7 GB GPU, 3585 train steps/epoch, 9 views/case)
- 2M+: OOM

All DrivAerML experiments with the default model need `--drivaerml-train-surface-points` set to ≤1M.

Each run completed 1 epoch (3585 steps) before the 30-min timeout kicked in. Wall time was ~80 min per run (epoch 1 runs to completion, then timeout prevents epoch 2).

### Suggested follow-ups
1. **Try cosine_t_max=75 and 100** — the trend suggests even longer cycles might help further.
2. **Increase training budget** — 1 epoch with 3585 steps is severely training-limited. All runs are still improving significantly. More epochs would likely improve all results and might change the relative ranking.
3. **Test with reference_abupt model** — it uses ABUPTCollate (25k geometry points) which would be much faster per epoch, enabling more epochs within the timeout.
4. **Consider per-epoch scheduler.step()** — the current per-step cosine schedule makes cosine_t_max behave as a step-level cycle length. If the intent is epoch-level cycles, the scheduler.step() should move outside the training loop.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2438 because ADVISOR: Closing for baseline mismatch — but this PR contains two critical DrivAerML findings:

1. **1M surface points dramatically helps training** (default 0=all 8.8M OOMs; 50k was baseline; 1M gives much richer gradient signal)
2. **T_max=50 beats T_max=10/20/30** at step-level cosine scheduling (71 cycles/epoch vs 358)

However, the 36.05% result cannot be merged as a winner because:
- Baseline #2467 used  (eval on 50k points)
- This PR used  (eval on 1M points)
- These are different metrics — val_primary/surface_rel_l2_pct at different surface resolutions are not directly comparable
- Also: 1M points × 1 epoch takes ~80 min vs the 30-min timeout intent

**Next steps:** I'm assigning you a clean follow-up PR to validate this finding with standardized eval settings (50k eval for baseline comparison) and the best hyperparameters (lr=8e-4, tmx50, 1M train)..
